pkg/metrics/prometheus.go

```diff
@@ -0,0 +1,152 @@
+package metrics
+
+import (
+	"github.com/grafana/grafana/pkg/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"net/http"
+	"strings"
+)
+
+const (
+	namespace = "grafana"
+	subsystem = "backend"
+)
+
+type PrometheusPublisher struct {
+	CounterCollectors map[string]*prometheus.CounterVec
+	GaugeCollectors   map[string]*prometheus.GaugeVec
+}
+
+func CreatePrometheusPublisher() (*PrometheusPublisher, error) {
+	go func() {
+		http.Handle("/metrics", prometheus.Handler())
+		err := http.ListenAndServe(":9200", nil)
```

Comment:
> That's interesting. Did not think Prometheus could do this accurately (or any TSDB). 

Noone can do it 100% accurately, you need logs for that. Prometheus should be a little bit more accurate though as we do a small bit of extrapolation.

> So if had a counter for user logins recorded every second (scraped or pushed doesnt matter). How would a query in prometheus look like if I wanted to get the total logins per day (or week) for the last year be? 

Each day is `increase(logins_total[1d])` and you'd use the query_range API with the right parameters to get that once per day. 

> can you sum a rate over time?

I'm not sure what you're asking here. I think you're looking to take the rate over a longer time period.

> Would not peeks be lost or maybe depends on the step parameter (how Prometheus process queries are still so strange to me)?

We're talking averages here over long time periods so peaks don't matter, unless they happen to be right at the start/end of the period.

For more general graphing you need to be careful that your rates are over a period longer than the step, otherwise you're dropping data.

> The problem then becomes if you need a TSDB that does roll ups, rolling up a raw counter using last (which is the most common I find) is very in accurate as server restarts will cause roll up to lose a lot of data. 

That's definitely the case. The Prometheus functions can tolerate one restart if the timing is right.

> The other alternative is to roll up using a derivative / rate function but then the aggregated data becomes a delta and the unaggregated becomes raw which makes it impossible to have the same query target both.

That's what you'd usually end up doing with Prometheus for the key stats.

> This is news to me. How can summing count deltas be inaccurate?
￼ 
It's more a question of exactly what the deltas represent.

Let's say you've got a delta every minute, and they happen to be coming in at 20 seconds past the minute. The data point in each minute slice does not tell you the amount in that minute, rather it's (on average) 66% of the previous minute and 33% this minute.

Thus if you were to do say a sum of an hours data, then you'd be missing 40 seconds of events and also including 40 seconds of events that you shouldn't be. Thus it is inaccurate.